Switch app to Spring Boot and internalize (lesson3)

KorolkovRS opened the pull request:
# Future
- Switch to Spring Boot
- Internilize output and questions file

# Changes
- issues from #2 done
- switch to Spring Boot
- application.properties swich to .yaml
- add internalize output and questions file
- add unit tests

KorolkovRS committed:
refactor: switch app to Spring Boot

KorolkovRS committed:
feature: add internalize



i18l

I18nService

i18l done

KorolkovRS committed:
tests: add unit tests

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/i18n_util/Internalizer.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public String internalizeMessage(String s) {
+        return ResourceBundle.getBundle(baseName, localeContext.getLocale()).getString(s);
```
А чем не понравился `MessageSource`? Он вроде по удобнее. Его можно внедрить и положить в поле. Опять же форматированные сообщения...

KorolkovRS replied:
Сделано в 4ed60ce

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/i18n_util/SystemLocaleResolver.java:
```diff
@@ -0,0 +1,14 @@
+public class SystemLocaleResolver implements LocaleContext {
+    @Override
+    public Locale getLocale() {
+        return Locale.getDefault();
```
Лучше из настроек взять

KorolkovRS replied:
Сделано в 4ed60ce

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/service/QuestionService.java:
```diff
@@ -0,0 +1,10 @@
+import java.util.List;
+import java.util.Locale;
+
+public interface QuestionService {
+    List<Question> getAllWithLocale(Locale locale);
+}
```
Справедливо и для сервиса и для дао (и ниже). Дао не нужно ничего знать ни про какую локаль. Ему нужно только имя файла для загрузки. Т.е. ему нужен внедренный провайдер файла, который внутри себя может знать про локаль и как из нее сделать имя файла. Всем остальным (`QuestionService` + `QuestionDao` + `ResourceProvider` + ...) локаль не нужна ни внутри, ни (тем более) на уровне интерфейсов

KorolkovRS replied:
Сделано в 4ed60ce

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/service/TestServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+
+    private int checkUserAnswer(Question q) {
+        try {
+            ioService.out(internalizer.internalizeMessage("testService.enter_answer") + q.getAnswers().size());
```
Можно сделать фасад для `ioService + internalizer`, чтобы сразу по ключу печатать

KorolkovRS replied:
Сделано в 2a81243

KorolkovRS committed:
refactor: notes to PR#3

refactor: ResourceBundle switch to MessageSource in Internalizer.java
refactor: Locale info remove from service and dao
feature: add PropertiesFileLocaleResolver

WARNING: check the tests where the local used
WARNING: SystemLocaleresolver depricated

KorolkovRS committed:
test: update tests

InternalizerTest, PropertiesFileLocaleResolverTest, ResourcePathResolverTest, ResourcePathProviderTest update

KorolkovRS committed:
feature: add IOServiceInternatiolizePolicy

Add IOServiceInternatiolizePolicy, update IOService, IOServiceImplm TestServiceIpl

KorolkovRS committed:
test: fix PropertiesFileLocaleResolverTest.java

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public User getUser() {
+        User user = new User();
+        ioService.out(internalizer.internalizeMessage("userService.enter_name"));
```
`IOServiceInternatiolizePolicy`?

KorolkovRS replied:
Исправлено в 3ccd794

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/policy/IOServiceInternatiolizePolicy.java:
```diff
@@ -0,0 +1,25 @@
+@Component
+@RequiredArgsConstructor
+public class IOServiceInternatiolizePolicy {
+    private final IOServiceImpl ioService;
```
Лучше по интерфейсу

KorolkovRS replied:
Исправлено в 3ccd794

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/policy/IOServiceInternatiolizePolicy.java:
```diff
@@ -0,0 +1,25 @@
+import ru.korolkovrs.spring.i18n_util.Internalizer;
+import ru.korolkovrs.spring.service.IOServiceImpl;
+
+@Component
+@RequiredArgsConstructor
+public class IOServiceInternatiolizePolicy {
```
И вроде это сервис?

KorolkovRS replied:
Исправлено в 3ccd794

KorolkovRS replied:
Если эта штука переезжает в сервисы, то кажется логичным связать ее с интерфейсом `IOService`, например:
public interface IOInternatiolizeService extends IOService {
//методы IOService 
void outWithInternalize(String s, Object... args);
}
Правда из-за этого придется как-то объяснять спрингу куда что инжектить, например через @Qualifier. На сколько это здесь применимо?

stvort replied:
Не логично. Сервис ввода-вывода это сервис ввода-вывода. А тут это сервис локализованного ввода-вывода

stvort commented on spring-03/src/main/java/ru/korolkovrs/spring/provider/ResourceProviderImpl.java:
```diff
@@ -0,0 +1,24 @@
+@Component
+@RequiredArgsConstructor
+public class ResourceProviderImpl implements ResourceProvider {
+    private final ResourcePathResolver pathResolver;
```
Опять же. Зачем `ResourceProvider`-у нужны `getCodes()`/`setCodes()` ? Ему нужно только имя файла. Лучше скрыть детали за интерфейсом

KorolkovRS replied:
Я так понимаю речь про `ResourcePathResolver`?
Там сеттер нужен для инжекта кодов из пропертей @ConfigurationProperties, а геттер по сути только для удобства тестирования:
   ``@Test
    void setCodesCorrect() {
        pathResolver.setCodes(codes);
        assertEquals(codes, pathResolver.getCodes());
    }``

stvort replied:
Да, но это лишняя информация. Можно класс оставить тем же самым, но сделать, чтобы он реализовывал интерфейс с одним методом и по этому интерфейсу сюда внедрить

KorolkovRS committed:
refactor: notes to PR#3

KorolkovRS committed:
refactor: ResourcePathResolver implements PropertiesFileResourcePathResolver interface

KorolkovRS commented:
Add internationalization
